### PR TITLE
Implemented BLE for IPhone and android can use bluetooth

### PR DESCRIPTION
**Pull request by GachiRoss**
replacing bluetooth classic with BLE

### CODE REVIEW

File: src/main/main.c [hunk 1]
```diff
@@ -240,68 +401,107 @@ void esp_bt_gap_cb(esp_bt_gap_cb_event_t event, esp_bt_gap_cb_param_t *param) {
...
+        int idx;
+        for (idx = 0; idx < PROFILE_NUM; idx++) {
+            /* ESP_GATT_IF_NONE, not specify a certain gatt_if, need to call every profile cb function */
+            if (gatts_if == ESP_GATT_IF_NONE || gatts_if == heart_rate_profile_tab[idx].gatts_if) {
+                if (heart_rate_profile_tab[idx].gatts_cb) {
+                    heart_rate_profile_tab[idx].gatts_cb(event, gatts_if, param);
+                }
+            }
+        }
+    } while (0);
```
Comment (SSpedsbjerg) [inline src/main/main.c:533]: what is the reason for running a while loop only once?

File: src/main/main.c [hunk 2]
```diff
@@ -171,67 +264,135 @@ void listen_to_buttons (void *pvParameters) {
...
+static void gap_event_handler(esp_gap_ble_cb_event_t event, esp_ble_gap_cb_param_t *param)
+{
     switch (event) {
-        case ESP_BT_GAP_AUTH_CMPL_EVT: {
-            //printf("Authentication status: %d, name: %s bda:[%s]", param->auth_cmpl.stat, param->auth_cmpl.device_name, bda2str(param->auth_cmpl.bda, bda_str, sizeof(bda_str)));
-            if (param->auth_cmpl.stat == ESP_BT_STATUS_SUCCESS) {
-                memcpy(trusted_device, param->auth_cmpl.bda, sizeof(esp_bd_addr_t));
+        case ESP_GAP_BLE_ADV_DATA_SET_COMPLETE_EVT:
+            adv_config_done &= (~ADV_CONFIG_FLAG);
+            if (adv_config_done == 0){
```
Comment (SSpedsbjerg) [inline src/main/main.c:374]: geussing this refence a enum, could that be clear so it is readable, cause I don't know what 0 means

File: src/main/main.c [hunk 3]
```diff
@@ -240,68 +401,107 @@ void esp_bt_gap_cb(esp_bt_gap_cb_event_t event, esp_bt_gap_cb_param_t *param) {
...
+            break;
     }
 }
 
-void ping_device(void *pvParameters) {
-    while (1) {
-        if (!paired) {
-            vTaskDelay(1000);
-            continue;
+static void gatts_event_handler(esp_gatts_cb_event_t event, esp_gatt_if_t gatts_if, esp_ble_gatts_cb_param_t *param)
```
Comment (SSpedsbjerg) [inline src/main/main.c:510]: Reason for this method?

File: src/main/main.c [hunk 4]
```diff
@@ -15,13 +15,39 @@
...
+
+#define BLE_TAG "SMART_DOORLOCK"
+
+#define PROFILE_NUM                 1
+#define PROFILE_APP_IDX             0
+#define ESP_APP_ID                  0x55
+#define SAMPLE_DEVICE_NAME          "SMART_DOORLOCK"
+#define SVC_INST_ID                 0
+
+#define ADV_CONFIG_FLAG             (1 << 0)
```
Comment (SSpedsbjerg) [inline src/main/main.c:49]: What is the reason for making bit-shifts here?

File: src/main/main.c [hunk 5]
```diff
@@ -15,13 +15,39 @@
...
+#define SCAN_RSP_CONFIG_FLAG        (1 << 1)
```
Comment (SSpedsbjerg) [inline src/main/main.c:50]: bit-shift

File: src/main/main.c [hunk 6]
```diff
@@ -49,6 +75,73 @@ bool open_button = false;
 bool lock_button = false;
 bool bt_connected = false;
 
+static uint8_t adv_config_done       = 0;
+
+static uint8_t service_uuid[16] = {
+    /* LSB <--------------------------------------------------------------------------------> MSB */
+    //first uuid, 16bit, [12],[13] is the value
+    0xfb, 0x34, 0x9b, 0x5f, 0x80, 0x00, 0x00, 0x80, 0x00, 0x10, 0x00, 0x00, 0xFF, 0x00, 0x00, 0x00,
```
Comment (SSpedsbjerg) [inline src/main/main.c:84]: pre-declared ID? these values seem random

**Review by SSpedsbjerg — CHANGES_REQUESTED** [review]

**Review by vimian — APPROVED** [review]

**Pull request merged**